ast/compile.go

```diff
@@ -3146,7 +3146,9 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
 	case Ref:
 		if RootDocumentRefs.Contains(term) {
 			if gv, ok := stack.Declared(v[0].Value.(Var)); ok {
-				term.Value = gv
+				if gv.IsGenerated() {
```

Comment:
Nice find!

I think it would be slightly better if instead of just checking if the var is generated (which fixes the issue but obscures the underlying issue) we check if the var was declared _explicitly_, i.e., with `:=` or `some` (or, in the future, as a function arg or other construct like `every`.) To do this we could check if the occurrence in the local var stack indicates the var is NOT a `seenVar` (or `newVar` which is the empty value.) This should be future proof.

While reviewing the related code, I noticed the `Occurence()` function on the stack is only intended to work in the local socpe. We'll need to check the local and global scopes in this case (e.g., if you have a call like this inside a comprehension it should "see" the local var declared outside.) Something like:

```golang
func (s localDeclaredVars) GlobalOccurrence(x Var) (varOccurrence, bool) {
  for i := len(s.vars)-1; i >= 0; i-- {
     if occ, ok := s.vars[i].occurrence[x]; ok {
      return occ, true
     }
   }
  return newVar, false
}
```

And then we can use this instead of checking if the var is generated:

```golang
x := v[0].Value(Var)
if occ, ok := stack.GlobalOccurence(x); ok && occ != seenVar {
  gv, _ := stack.Declared(v[0].Value.(Var))
  term.Value = gv
}
```

LMKWYT.

Reply:
Updated the check logic.